pages/posts/build-oci-image-from-scratch.md

```diff
@@ -0,0 +1,391 @@
+{
+    "schemaVersion": 2,
+    "manifests": [
+        {
+            "mediaType": "application/vnd.oci.image.manifest.v1+json",
+--------->  "digest": "sha256:8289bd1bdc2a1fae2d2d717b7c40baaedc4c5c4d9c9f4f1a1b045287067e9f2c",
+--------->  "size": 748,
+            "annotations": {
+                "org.opencontainers.image.ref.name": "hello:alpine1"
+            }
+        }
+    ]
+}
+```
+
+If we pack and load the image again with podman, we can see the modified image in action:
+```
+$ tree
+├── blobs
+│   └── sha256
+│       ├── 75b148a9a5b61403d082f1930ccc779bfdac5acddb48ed8f90ccdc4219d51268
+│       ├── 8289bd1bdc2a1fae2d2d717b7c40baaedc4c5c4d9c9f4f1a1b045287067e9f2c
+│       ├── c37c06cdec9d6a0f2a2d55deb5aa002b26b37b17c02c2eca908fc062af5f53eb
+│       └── c59d5203bc6b8b6ef81f3f6b63e32c28d6e47be806ba8528f8766a4ca506c7ba
+└── index.json
+$ tar -cvf hello-alpine.tar *
+
+$ podman load < hello-alpine.tar
+Getting image source signatures
+Copying blob c59d5203bc6b done   |
+Copying blob c37c06cdec9d done   |
+Copying config 75b148a9a5 done   |
+Writing manifest to image destination
+Loaded image: localhost/hello:alpine
+
+$ podman run --rm localhost/hello:alpine reader
+Hello, reader!
+real    0m 0.00s
+user    0m 0.00s
+sys     0m 0.00s
+
+$ podman image ls hello
+REPOSITORY   TAG       IMAGE ID       CREATED   SIZE
+hello        scratch   25e8b3bd9720   N/A       3.67MB
+hello        alpine    3d0268e9a91e   N/A       11MB
+```
+
+So there we have it, our two images loaded into the docker daemon. We can run both of them and see the same output from both of them. One interesting thing to note is the size difference between the two images. Alpine, while still having a lower footprint is larger than scratch, but nothing extraordinary. Hopefully this example demonstrated interaction between different layers in an image and how a manifest file can represent multiple images from a single file.
+
+
+# Conclusion
+I was initially planning to title this article "Container Image Internals" and the first draft of that came out all theoretical and while reading it, even I felt bored even though I enjoyed finding about all I could about container images for the post. But then I decided to document the fun part of this whole journey and here we are. In this article, we saw how we can "build" a docker image from scratch. But that's not to say that we'd be doing something like that on a day to day basis. You should never find yourself doing this as part of your regular development workflow. This article intentionally uses a simple example and so it misses out on important nuances such as how layers use an impelementation of UnionFS and don't pack simplified tarballs anymore, etc.
+
+But alas, it's always fun to crack open something and to understanding things so I hope this article helped demistify container images to a certain extent. If you find anything that seems wrong or if you think there's something that can be improved, feel free to reach out.
```

Comment:
Keep it tight and simple.

1. The examples used in this articles is not something you'd do in your daily work. Their purpose is to demonstrate of inner working of [insert what inner workings].
2. Reach out.
3. Done!